### PR TITLE
GlowTTS model

### PR DESCRIPTION
**Pull request by sam1373**
Adds GlowTTS model code to the tts collection. Also adds some corresponding neural types and jenkins test.

### CODE REVIEW

File: Jenkinsfile [hunk 1]
```diff
@@ -526,6 +526,19 @@ pipeline {
...
+            train_dataset=/home/TestData/an4_dataset/an4_train.json \
+            validation_datasets=/home/TestData/an4_dataset/an4_val.json \
+            trainer.gpus="[1]" \
+            +trainer.fast_dev_run=True \
+            trainer.distributed_backend=null \
+            trainer.max_epochs=-1 \
+            model.train_ds.batch_size=4 \
+            model.validation_ds.batch_size=4'
+          }
+        }
```
Comment (blisc) [inline Jenkinsfile:541]: Can you move this block into a separate stage? Make a L2: TTS Fast Dev Runs 1 and a L2: TTS Fast Dev Runs 2

File: nemo/collections/tts/data/datalayers.py [hunk 2]
```diff
@@ -117,3 +124,52 @@ def __getitem__(self, index):
 
     def __len__(self):
         return len(self.collection)
+
+
+@experimental
+class AudioToPhonemesDataset(_AudioTextDataset):
```
Comment (blisc) [inline nemo/collections/tts/data/datalayers.py:130]: Can you docstring this class?

File: nemo/collections/tts/losses/glow_tts_loss.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    @typecheck()
+    def forward(self, z, y_m, y_logs, logdet, logw, logw_, x_lengths, y_lengths):
+
+        logdet = torch.sum(logdet)
+        l_mle = 0.5 * math.log(2 * math.pi) + (
+            torch.sum(y_logs) + 0.5 * torch.sum(torch.exp(-2 * y_logs) * (z - y_m) ** 2) - logdet
+        ) / (torch.sum(y_lengths) * z.shape[1])
+
+        l_length = torch.sum((logw - logw_) ** 2) / torch.sum(x_lengths)
+        return l_mle, l_length, logdet
```
Comment (blisc) [inline nemo/collections/tts/losses/glow_tts_loss.py:71]: Any reason you return 3 variables here as opposed to a singular loss?

File: nemo/collections/tts/modules/glow_tts.py [hunk 4]
```diff
@@ -0,0 +1,419 @@
...
+        filter_channels,
+        filter_channels_dp,
+        n_heads,
+        n_layers,
+        kernel_size,
+        p_dropout,
+        window_size,
+        mean_only=False,
+        prenet=False,
+        gin_channels=0,
```
Comment (blisc) [inline nemo/collections/tts/modules/glow_tts.py:43]: Can you typehint these?

File: nemo/collections/tts/modules/glow_tts.py [hunk 5]
```diff
@@ -0,0 +1,419 @@
...
+        kernel_size,
+        dilation_rate,
+        n_blocks,
+        n_layers,
+        p_dropout=0.0,
+        n_split=4,
+        n_sqz=2,
+        sigmoid_scale=False,
+        gin_channels=0,
+    ):
```
Comment (blisc) [inline nemo/collections/tts/modules/glow_tts.py:187]: Typehints please

File: nemo/collections/tts/modules/glow_tts.py [hunk 6]
```diff
@@ -0,0 +1,419 @@
...
+        pass
+
+    @classmethod
+    def restore_from(cls, restore_path: str):
+        """TODO: Implement"""
+        pass
+
+
+class GlowTTSModule(NeuralModule):
+    def __init__(self, encoder_module, decoder_module, n_speakers=1, gin_channels=0):
```
Comment (blisc) [inline nemo/collections/tts/modules/glow_tts.py:304]: Typehints please

File: nemo/collections/tts/parts/glow_tts_parser.py [hunk 7]
```diff
@@ -0,0 +1,381 @@
+import re
```
Comment (blisc) [inline nemo/collections/tts/parts/glow_tts_parser.py:1]: Add copyright header

Comment (blisc) [inline nemo/collections/tts/parts/glow_tts_parser.py:1]: Maybe move this file to helpers instead of parts?

File: nemo/collections/tts/parts/glow_tts.py [hunk 8]
```diff
@@ -0,0 +1,606 @@
+import math
```
Comment (blisc) [inline nemo/collections/tts/parts/glow_tts.py:1]: Copyright header

Comment (blisc) [inline nemo/collections/tts/parts/glow_tts.py:1]: I would consider splitting this file partly into helpers and partly into modules/submodules.py or another file inside of modules

File: nemo/collections/tts/modules/glow_tts.py [hunk 9]
```diff
@@ -0,0 +1,419 @@
+import math
```
Comment (blisc) [inline nemo/collections/tts/modules/glow_tts.py:1]: Copyright header.

File: nemo/collections/tts/parts/glow_tts.py (hunk 8 above)
Comment (sam1373) [reply nemo/collections/tts/parts/glow_tts.py:1]: I thought that helpers was more for functions separate from the actual model, while "parts" in other collections seems to be used for submodules and functions for them? E.g. in asr parts folder contains both the parsers and the jasper submodules.

File: nemo/collections/tts/losses/glow_tts_loss.py (hunk 3 above)
Comment (sam1373) [reply nemo/collections/tts/losses/glow_tts_loss.py:71]: I am returning them like this mostly because it can be useful to have these values separately for monitoring purposes.

File: nemo/collections/tts/parts/glow_tts.py (hunk 8 above)
Comment (blisc) [reply nemo/collections/tts/parts/glow_tts.py:37]: I tried to have torch.nn.modules under modules inside TTS and all other "python" functions inside helpers

Comment (sam1373) [reply nemo/collections/tts/parts/glow_tts.py]: I changed this to be a file in modules, but I don't think it makes sense to put functions that are directly relevant to the model in helpers

Comment (blisc) [reply nemo/collections/tts/parts/glow_tts.py]: Hmm, I have a bunch of model specific code inside of helpers. Should we rename helpers to parts?
The directory structure could definitely be improved if you have suggestions to make.

File: nemo/collections/tts/data/datalayers.py [hunk 10]
```diff
@@ -117,3 +124,73 @@ def __getitem__(self, index):
...
+            "transcript_length": NeuralType(tuple("B"), LengthsType()),
+        }
+
+    def __init__(
+        self,
+        manifest_filepath: str,
+        cmu_dict_path: str,
+        sample_rate: int,
+        int_values: bool = False,
+        augmentor: "nemo.collections.asr.parts.perturb.AudioAugmentor" = None,
```
Comment (stasbel) [inline nemo/collections/tts/data/datalayers.py:153]: 1) Can you justify need for str type marking here?
2) [Here and below] Why do you type args in both docstring and method declaration? For example, this is strongly discourage in Google's style guide.

File: nemo/collections/tts/modules/glow_tts_parser.py [hunk 11]
```diff
@@ -0,0 +1,417 @@
...
+import inflect
+from unidecode import unidecode
+
+
+class CMUDict:
+    """Thin wrapper around CMUDict data. http://www.speech.cs.cmu.edu/cgi-bin/cmudict"""
+
+    def __init__(self, file_or_path, keep_ambiguous=True):
+
+        self.valid_symbols = [
```
Comment (stasbel) [inline nemo/collections/tts/modules/glow_tts_parser.py:49]: https://stackoverflow.com/a/58584557/6926596

File: nemo/collections/tts/data/datalayers.py (hunk 10 above)
Comment (sam1373) [reply nemo/collections/tts/data/datalayers.py]: 1. Changed it, though this is how it is in the asr collection datalayers
2. Changed to method

File: nemo/collections/tts/modules/glow_tts_parser.py (hunk 11 above)
Comment (sam1373) [reply nemo/collections/tts/modules/glow_tts_parser.py:54]: fixed

**Comment by sam1373** [comment]
SWIPAT has been approved

File: nemo/collections/tts/losses/glow_tts_loss.py [hunk 12]
```diff
@@ -0,0 +1,93 @@
+# MIT License
```
Comment (blisc) [inline nemo/collections/tts/losses/glow_tts_loss.py:1]: Can we follow this: https://github.com/NVIDIA/NeMo/pull/988#discussion_r465936184 for copyright headers?

File: nemo/collections/tts/modules/glow_tts.py [hunk 13]
```diff
@@ -0,0 +1,458 @@
+# MIT License
```
Comment (blisc) [inline nemo/collections/tts/modules/glow_tts.py:1]: Can we follow this: https://github.com/NVIDIA/NeMo/pull/988#discussion_r465936184 for copyright headers?

File: nemo/collections/tts/modules/glow_tts_parser.py [hunk 14]
```diff
@@ -0,0 +1,347 @@
+# MIT License
```
Comment (blisc) [inline nemo/collections/tts/modules/glow_tts_parser.py:1]: Can we follow this: https://github.com/NVIDIA/NeMo/pull/988#discussion_r465936184 for copyright headers?

File: nemo/collections/tts/modules/glow_tts_submodules.py [hunk 15]
```diff
@@ -0,0 +1,643 @@
+# MIT License
```
Comment (blisc) [inline nemo/collections/tts/modules/glow_tts_submodules.py:1]: Can we follow this: https://github.com/NVIDIA/NeMo/pull/988#discussion_r465936184 for copyright headers?
